contrib/infiniband/keys.ic

```diff
@@ -58,20 +62,59 @@ uint32_t update_rkey_restart(uint32_t rkey, int pd_id)
 
 void update_rkey_send(struct ibv_send_wr * wr, int pd_id)
 {
-  if (is_restart) {
-    if (wr == NULL) {
-      return;
-    }
-    if (wr->opcode == IBV_WR_RDMA_WRITE || 
-        wr->opcode == IBV_WR_RDMA_WRITE_WITH_IMM || 
-        wr->opcode == IBV_WR_RDMA_READ) {
-      wr->wr.rdma.rkey = update_rkey_restart(wr->wr.rdma.rkey, pd_id);
-    }
-    
-    if (wr->opcode == IBV_WR_ATOMIC_CMP_AND_SWP ||
-        wr->opcode == IBV_WR_ATOMIC_FETCH_AND_ADD) {
-      wr->wr.atomic.rkey = update_rkey_restart(wr->wr.atomic.rkey, pd_id);
-    }
-    update_rkey_send(wr->next, pd_id);
+  if (wr == NULL) {
+    return;
+  }
+
+  if (wr->opcode == IBV_WR_RDMA_WRITE ||
+      wr->opcode == IBV_WR_RDMA_WRITE_WITH_IMM ||
+      wr->opcode == IBV_WR_RDMA_READ) {
+    wr->wr.rdma.rkey = update_rkey_restart(wr->wr.rdma.rkey, pd_id);
+  }
+
+  if (wr->opcode == IBV_WR_ATOMIC_CMP_AND_SWP ||
+      wr->opcode == IBV_WR_ATOMIC_FETCH_AND_ADD) {
+    wr->wr.atomic.rkey = update_rkey_restart(wr->wr.atomic.rkey, pd_id);
+  }
+  update_rkey_send(wr->next, pd_id);
+}
+
+void update_ud_send_restart(struct ibv_send_wr *wr) {
+  struct internal_ibv_ah *internal_ah;
+  uint32_t size;
+  ibv_ud_qp_id orig_id, curr_id;
+
+  if (wr == NULL) {
+    return;
   }
+
+  internal_ah = ibv_ah_to_internal(wr->wr.ud.ah);
+  orig_id.qpn = wr->wr.ud.remote_qpn;
+  orig_id.lid = internal_ah->attr.dlid;
+
+  dmtcp_send_query_to_coordinator("qp_info",
+                                  &orig_id, sizeof(orig_id),
+                                  &curr_id, &size);
+
+  if (size == sizeof(ibv_ud_qp_id)) {
+    // The qp is created before checkpoint, qp_num changes
+    wr->wr.ud.remote_qpn = curr_id.qpn;
+  }
```

Comment:
The "if" statement tests `if (size == sizeof(ibv_ud_qp_id))`.  This code eems unusual and difficult to maintain.  It seems to be saying that the `curr_id` of size `size` that was returned by a query could be of more than one type.  So, you're testing the _size_ of the data structure that was returned, to decide if it is the right type.

Normally, I would expect the first field of a struct to be a constant of a certain type.  So, I would do something like `if (curr_id.type == UD_QP_ID) { ... }`.  Then, anyone maintaining this code can debug it more easily by looking at the field `curr_id.type`.

It's also possible that I didn't understand this code correctly.  But in that case, the right thing to do would be to add an extra comment before the "if" statement, to help people like me to understand this code.